src/main/java/com/conveyal/r5/profile/FastRaptorWorker.java

```diff
@@ -855,16 +861,17 @@ private void doTransfers (RaptorState state) {
      * Find all patterns that could lead to improvements in the next round after the given state's raptor round.
      * Specifically, the patterns passing through all stops that were updated in the given state's round.
      * The pattern indexes returned are limited to those in the supplied set.
+     * @param withinMinute EXPLAIN
```

Comment:
Is this a correct explanation? Do we need to distinguish between departure minutes and iterations within those minutes?

```suggestion
     * Find all patterns that could lead to improvements in the next round after the given state's raptor round.
     * Specifically, the patterns passing through all stops that were updated in the given state's round. 
     * For frequency upper-bound and schedule searches, no previously checked (i.e. later) departure time can yield 
     * earlier arrival times at stops, so only patterns passing through stops updated in this round need to be checked in the next round. 
     * For frequency searches, where the randomized schedules vary, a previously checked departure minute could lead to an
     * earlier arrival, so the incumbent best arrival time at each stop needs to be compared directly against the arrival time in this round.
     * The pattern indexes returned are limited to those in the supplied set.
     * @param withinMinute if true, consider updates only within this round; should be false when previously checked rounds (i.e. later departure minutes) can have earlier arrivals at stops (e.g. in the case of frequency searches with random schedules)
```

Reply:
Thanks for supplying this draft explanation. I will need to put some thought into a clear explanation of what we're doing here, because it's potentially confusing and critical for correctness. 

The facts you mention here are very relevant, but the full line of reasoning is more indirect. The fact that for randomized frequency routes, a later departure minute may yield an earlier arrival (due to a more favorable random schedule) is why we apply each set of randomized frequency schedules on its own copy of the state for a departure minute, then throw it away after extracting the travel times. The resulting state is simply not reusable because it's reflecting a network-wide (scheduled+frequency) timetable that is never used again outside that one iteration. Range-raptor is applied to a certain subset of routes, but final travel times are extracted from raptor runs for a superset of those routes that are dead-end "side branches" off that main range-raptor chain of departure minutes.

This fractional application of range-raptor to only a subset of routes is what necessitates two different ways of tracking updated stops. Range-raptor speeds up the search process by skipping all the routes where we could not improve on the result from a later departure minute, and using certain facts about causality and time it cascades that route-skipping down through the rounds. One of those facts is that if in round N we can reuse an optimal stop arrival time from a later departure minute, all rides in round N+1 boarding at that stop are already optimal and don't need to be reconsidered. This is dependent on the fact that the state from the later departure minute was computed for exactly the same set of routes as the current departure minute. When we add more routes (here, the randomized frequency routes) this cascading optimality is no longer guaranteed. It is conditional on the already-optimal stop having no newly added routes passing through it.

In the worst case where there are tons of frequency routes, I suspect this optimization of layering them on top of the scheduled range raptor results would not improve run time much or at all. The added frequency routes could invalidate all the cascading route-skipping and cause every scheduled and frequency route to be re-explored at every round of every iteration. In performance testing it would be interesting to go to extremes in switching off optimizations, and do each frequency draw in isolation, starting completely from scratch with fresh empty state. I think we'd discover this set of optimizations is only effective in certain cases, which happen to be fairly common (a small number of frequency routes layered on a complex scheduled network).

Reply:
The terminology here can get very tricky, because we've got departure minutes, rounds, and draws, and those can vary in different combinations. For example when saying "this round" it's not clear if we mean "only things updated in this round of this draw and this departure minute" or "this round as a state accumulated across all departure minutes". Eventually our documentation will need diagrams for all this, with a single unambiguous name for each concept. I will commit your explanation and then revise it in a follow-up pull request.